tensorbay/opendataset/nuScenes/loader.py

```diff
@@ -0,0 +1,279 @@
+            )
+            if "test" not in subset:
+                frame["LIDAR_TOP"].label.box3d = _load_labels(
+                    current_frame,
+                    segment.sensors["LIDAR_TOP"].extrinsics,
+                    lidar_ego_pose_token,
+                    annotation_files,
+                )
+
+            segment.append(frame)
+            current_frame = annotation_files["samples"][current_frame]["next"]
+        yield segment
+
+
+def _load_annotation_files(info_path: str) -> Dict[str, Any]:
+    annotation_files = {
+        "samples": _load_info(os.path.join(info_path, "sample.json")),
+        "frame_data": _load_sample_info(os.path.join(info_path, "sample_data.json")),
+        "calibrated_sensors": _load_info(os.path.join(info_path, "calibrated_sensor.json")),
+        "ego_poses": _load_info(os.path.join(info_path, "ego_pose.json")),
+        "sensor": _load_info(os.path.join(info_path, "sensor.json")),
+    }
+    with open(os.path.join(info_path, "scene.json"), "r") as file:
+        annotation_files["scenes"] = json.load(file)
+    if "test" not in info_path:
+        annotation_files["sample_annotations"] = _load_sample_info(
+            os.path.join(info_path, "sample_annotation.json"), no_key_frame=True
+        )
+        annotation_files["instance"] = _load_info(os.path.join(info_path, "instance.json"))
+        annotation_files["category"] = _load_info(os.path.join(info_path, "category.json"))
+        annotation_files["attribute"] = _load_info(os.path.join(info_path, "attribute.json"))
+        annotation_files["visibility"] = _load_info(os.path.join(info_path, "visibility.json"))
+
+    return annotation_files
+
+
+def _load_info(filepath: str) -> Dict[str, Any]:
+    with open(filepath, "r") as fp:
+        info = json.load(fp)
+    return {item["token"]: item for item in info}
+
+
+def _load_sample_info(filepath: str, no_key_frame: bool = False) -> Dict[str, List[Any]]:
+    with open(filepath, "r") as fp:
+        info = json.load(fp)
+    info_with_keys = defaultdict(list)
+    for item in info:
+        if no_key_frame or item["is_key_frame"]:
+            info_with_keys[item["sample_token"]].append(item)
+    return info_with_keys
+
+
+def _load_frame(
+    current_frame: str,
+    annotation_files: Dict[str, Any],
+    sensors: Sensors,
+    subset_path: str,
+) -> Tuple[Frame, str]:
+    frame = Frame()
+    lidar_ego_pose_token = ""
+    for sensor_frame in annotation_files["frame_data"][current_frame]:
+        calibrated_sensor_info = annotation_files["calibrated_sensors"][
+            sensor_frame["calibrated_sensor_token"]
+        ]
+        common_sensor = annotation_files["sensor"][calibrated_sensor_info["sensor_token"]]
+        sensor_name = common_sensor["channel"]
+        sensor_type = common_sensor["modality"]
+
+        if sensor_type == "lidar":
+            lidar_ego_pose_token = sensor_frame["ego_pose_token"]
+
+        if sensor_name not in sensors:
+            sensors.add(
```

Comment:
It will be better to make function constant, which means the function will not modify the input argument.

Reply:
Also it is weird to add sensors in a function named `_load_frame`